contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,758 @@
+  )
+    ConfirmedOwner(msg.sender)
+  {
+    LINK = LinkTokenInterface(link);
+    LINK_ETH_FEED = AggregatorV3Interface(linkEthFeed);
+    BLOCKHASH_STORE = BlockhashStoreInterface(blockhashStore);
+  }
+
+  function registerProvingKey(
+    address oracle,
+    uint256[2] calldata publicProvingKey
+  )
+    external
+    onlyOwner()
+  {
+    bytes32 kh = hashOfKey(publicProvingKey);
+    if (s_serviceAgreements[kh] != address(0)) {
+      revert KeyHashAlreadyRegistered(kh);
+    }
+    s_serviceAgreements[kh] = oracle;
+    emit NewServiceAgreement(kh, oracle);
+  }
+
+  /**
+   * @notice Returns the serviceAgreements key associated with this public key
+   * @param publicKey the key to return the address for
+   */
+  function hashOfKey(
+    uint256[2] memory publicKey
+  )
+    public
+    pure
+    returns (
+      bytes32
+    )
+  {
+    return keccak256(abi.encodePacked(publicKey));
+  }
+
+  function setConfig(
+    uint16 minimumRequestBlockConfirmations,
+    uint32 fulfillmentFlatFeeLinkPPM,
+    uint32 maxGasLimit,
+    uint32 stalenessSeconds,
+    uint32 gasAfterPaymentCalculation,
+    uint96 minimumSubscriptionBalance,
+    int256 fallbackWeiPerUnitLink
+  )
+    external
+    onlyOwner()
+  {
+    s_config = Config({
+      minimumRequestBlockConfirmations: minimumRequestBlockConfirmations,
+      fulfillmentFlatFeeLinkPPM: fulfillmentFlatFeeLinkPPM,
+      maxGasLimit: maxGasLimit,
+      stalenessSeconds: stalenessSeconds,
+      gasAfterPaymentCalculation: gasAfterPaymentCalculation,
+      minimumSubscriptionBalance: minimumSubscriptionBalance
+    });
+  s_fallbackWeiPerUnitLink = fallbackWeiPerUnitLink;
+  emit ConfigSet(
+      minimumRequestBlockConfirmations,
+      fulfillmentFlatFeeLinkPPM,
+      maxGasLimit,
+      stalenessSeconds,
+      gasAfterPaymentCalculation,
+      minimumSubscriptionBalance,
+      fallbackWeiPerUnitLink
+    );
+  }
+
+  /**
+   * @notice read the current configuration of the coordinator.
+   */
+  function getConfig()
+    external
+    view
+    returns (
+      uint16 minimumRequestBlockConfirmations,
+      uint32 fulfillmentFlatFeeLinkPPM,
+      uint32 maxGasLimit,
+      uint32 stalenessSeconds,
+      uint32 gasAfterPaymentCalculation,
+      uint96 minimumSubscriptionBalance,
+      int256 fallbackWeiPerUnitLink
+    )
+  {
+    Config memory config = s_config;
+    return (
+      config.minimumRequestBlockConfirmations,
+      config.fulfillmentFlatFeeLinkPPM,
+      config.maxGasLimit,
+      config.stalenessSeconds,
+      config.gasAfterPaymentCalculation,
+      config.minimumSubscriptionBalance,
+      s_fallbackWeiPerUnitLink
+    );
+  }
+
+  function recoverFunds(
+    address to
+  )
+    external
+    onlyOwner()
+  {
+    uint256 externalBalance = LINK.balanceOf(address(this));
+    uint256 internalBalance = uint256(s_totalBalance);
+    if (internalBalance > externalBalance) {
+      revert BalanceInvariantViolated(internalBalance, externalBalance);
```

Comment:
Assuming the contract is bugless it could never happen. I guess we could leave out the revert, but I thought it might be a useful alarm after which we could recoverFunds, kill the contract and figure out what went wrong